mining/src/monitor.rs

```diff
@@ -0,0 +1,94 @@
+use super::MiningCounters;
+use kaspa_core::{
+    info,
+    task::{
+        service::{AsyncService, AsyncServiceFuture},
+        tick::{TickReason, TickService},
+    },
+    trace,
+};
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+const MONITOR: &str = "mempool-monitor";
+
+pub struct MiningMonitor {
+    // Counters
+    counters: Arc<MiningCounters>,
+
+    // Tick service
+    tick_service: Arc<TickService>,
+}
+
+impl MiningMonitor {
+    pub fn new(counters: Arc<MiningCounters>, tick_service: Arc<TickService>) -> MiningMonitor {
+        MiningMonitor { counters, tick_service }
+    }
+
+    pub async fn worker(self: &Arc<MiningMonitor>) {
+        let mut last_snapshot = self.counters.snapshot();
+        let mut last_log_time = Instant::now();
+        let snapshot_interval = 10;
+        loop {
+            if let TickReason::Shutdown = self.tick_service.tick(Duration::from_secs(snapshot_interval)).await {
+                // Let the system print final logs before exiting
+                tokio::time::sleep(Duration::from_millis(500)).await;
+                break;
+            }
+
+            let snapshot = self.counters.snapshot();
+            if snapshot == last_snapshot {
+                // No update, avoid printing useless info
+                last_log_time = Instant::now();
+                continue;
+            }
+
+            // Subtract the snapshots
+            let delta = &snapshot - &last_snapshot;
+            let now = Instant::now();
+            let elapsed = (now - last_log_time).as_secs_f64();
+
+            info!("Processed {} unique transactions in the last {:.2}s ({:.2} avg txs/s, in: {} via RPC, {} via P2P, out: {} via accepted blocks, {:.2}% e-tps)",
+                delta.tx_accepted_counts,
+                elapsed,
+                delta.tx_accepted_counts as f64 / elapsed,
```

Comment:
it feels like the log line has some repeating redundant info and is a little too long. I think `Processed {} unique transactions in the last {:.2}s` can be removed, wdyt? `avg txs/s` already contains this info, and the consensus monitor already prints raw tx counts 

I suggest the line looks like: `[Unique tx stats] {:.2} avg txs/s, in: {} via RPC, {} via P2P, out: {} via accepted blocks, {:.2}% e-tps`

Reply:
Nice :)

But we still need the elapsed time somewhere. Consensus and Mining monitors are not synced. The log lines are switching order sometimes, their respective schedule may drift differently as the node runs for a long time and when grepping the log for this line, we loose the elapsed info altogether.